File: android/app/src/main/java/com/rclipper/app/render/RenderManifest.java
package com.rclipper.app.render;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * The server's render manifest, parsed into something the renderer can walk.
 *
 * This is the Java half of `src/lib/mobile/deviceRenderContract.ts`. It is
 * deliberately strict: a field the manifest is required to carry is read with
 * `getX` (which throws) rather than `optX` (which invents a default). A phone
 * that renders a manifest it half-understood produces a video that looks nearly
 * right, and "nearly right" is the failure mode nobody catches until a customer
 * does.
 *
 * Validation that has already happened server-side is NOT repeated here, with
 * one exception: anything that would make this code index out of bounds or
 * divide by zero is checked, because a malformed manifest must fail with a
 * message rather than a stack trace.
 */
public final class RenderManifest {

    public static final int SUPPORTED_VERSION = 4;

    public final int version;
    public final String attemptId;
    public final String jobId;
    public final String requestId;
    public final String step;
    /** "montage" | "master" | "final". */
    public final String stage;
    public final String ratio;
    public final int width;
    public final int height;
    public final int fps;
    public final double sceneTransitionSeconds;
    public final List<Source> sources;
    public final List<Scene> scenes;
    /** The montage for a master render, or the master for a final render. */
    public final String inputVideoUrl;
    public final String voiceUrl;
    public final String musicUrl;
    public final AudioSpec audio;
    public final List<Caption> captions;
    public final List<String> captionLanguages;
    public final Template template;
    public final boolean coverRequired;
    public final double coverAtSeconds;
    public final long maxOutputBytes;
    /**
     * Render this master or final straight from the approved originals — the
     * montage, the mix and (for a final) the template and captions in ONE
     * encode — instead of downloading the previous stage's export and encoding
     * it again. Sent only to builds that declare they can (plugin version 6+);
     * absent on older manifests, which keep the download path.
     */
    public final boolean buildFromSources;

    public static final class Source {
        public final String assetId;
        /** "image" | "clip". */
        public final String kind;
        /** Exactly one of url/localId is set. */
        public final String url;
        public final String localId;
        public final String mimeType;
        public final Double durationSeconds;

        Source(JSONObject json) throws JSONException {
            assetId = json.getString("assetId");
            kind = json.getString("kind");
            url = json.optString("url", null);
            localId = json.optString("localId", null);
            mimeType = json.getString("mimeType");
            durationSeconds = json.isNull("durationSeconds") ? null : json.getDouble("durationSeconds");
        }

        public boolean isImage() {
            return "image".equals(kind);
        }
    }

    public static final class Shot {
        public final String sourceAssetId;
        /** The slot this shot occupies before any dissolve overlap. */
        public final double durationSeconds;
        public final String motion;
        public final double trimStartSeconds;
        /** Null when the clip plays to its natural end. */
        public final Double trimEndSeconds;
        public final float focusX;
        public final float focusY;
        /** 0 = whole picture, 1 = fill the frame (the default). See ShotFraming. */
        public final float frameZoom;
        public final float playbackRate;

        Shot(JSONObject json) throws JSONException {
            sourceAssetId = json.getString("sourceAssetId");
            durationSeconds = json.getDouble("durationSeconds");
            motion = json.optString("motion", "static");
            trimStartSeconds = json.optDouble("trimStartSeconds", 0d);
            trimEndSeconds = json.has("trimEndSeconds") && !json.isNull("trimEndSeconds")
                ? json.getDouble("trimEndSeconds") : null;
            focusX = (float) json.optDouble("focusX", 0.5d);
            focusY = (float) json.optDouble("focusY", 0.5d);
            frameZoom = (float) Math.max(0d, Math.min(1d, json.optDouble("frameZoom", 1d)));
            playbackRate = (float) json.optDouble("playbackRate", 1d);
            if (durationSeconds <= 0) throw new JSONException("A shot has no duration");
        }
    }

    public static final class Scene {
        public final int sceneNumber;
        /** "cut" | "fade" | "slide" | "zoom". */
        public final String transitionIn;
        public final double shotTransitionSeconds;
        public final List<Shot> shots;

        Scene(JSONObject json) throws JSONException {
            sceneNumber = json.optInt("sceneNumber", 0);
            transitionIn = json.optString("transitionIn", "fade");
            shotTransitionSeconds = json.optDouble("shotTransitionSeconds", 0.2d);
            JSONArray array = json.getJSONArray("assets");
            if (array.length() == 0) throw new JSONException("A scene has no shots");
            shots = new ArrayList<>(array.length());
            for (int i = 0; i < array.length(); i++) shots.add(new Shot(array.getJSONObject(i)));
        }
    }

    /** Every number the mixer needs; see `deviceRenderAudio.ts` for the why. */
    public static final class AudioSpec {
        public final boolean musicSelected;
        public final int sampleRate;
        public final double leadInSeconds;
        public final float musicBedVolume;
        public final float duckRatio;
        public final float duckThreshold;
        public final float duckAttackMs;
        public final float duckReleaseMs;
        public final float targetLufs;
        public final float truePeakDb;
        public final float maxGainDb;
        public final float limit;
        public final Double voiceDurationSeconds;

        AudioSpec(JSONObject json) throws JSONException {
            musicSelected = json.getBoolean("musicSelected");
            sampleRate = json.getInt("sampleRate");
            leadInSeconds = json.getDouble("voiceLeadInSeconds");
            musicBedVolume = (float) json.getDouble("musicBedVolume");
            duckRatio = (float) json.getDouble("musicDuckRatio");
            duckThreshold = (float) json.getDouble("musicDuckThreshold");
            duckAttackMs = (float) json.getDouble("musicDuckAttackMs");
            duckReleaseMs = (float) json.getDouble("musicDuckReleaseMs");
            targetLufs = (float) json.getDouble("voiceTargetLufs");
            truePeakDb = (float) json.getDouble("voiceTruePeakDb");
            maxGainDb = (float) json.getDouble("voiceMaxGainDb");
            limit = (float) json.getDouble("mixLimit");
            voiceDurationSeconds = json.isNull("voiceDurationSeconds")
                ? null : json.getDouble("voiceDurationSeconds");
        }
    }

    public static final class Caption {
        public final double startSeconds;
        public final double endSeconds;
        public final String textThai;
        public final String textEnglish;
        public final String textChinese;

        Caption(JSONObject json) throws JSONException {
            startSeconds = json.getDouble("startSeconds");
            endSeconds = json.getDouble("endSeconds");
            textThai = json.optString("textThai", "");
            textEnglish = json.optString("textEnglish", "");
            textChinese = json.optString("textChinese", "");
        }

        public String textFor(String language) {
            switch (language) {
                case "th": return textThai;
                case "en": return textEnglish;
                case "zh": return textChinese;
                default: return "";
            }
        }
    }

    public static final class Template {
        public final String id;
        public final String frame;
        public final String canvas;
        public final List<String> decor;
        public final int primary;
        public final int secondary;
        public final int accent;
        public final int neutral;

        Template(JSONObject json) throws JSONException {
            id = json.getString("id");
            frame = json.getString("frame");
            canvas = json.getString("canvas");
            JSONArray decorArray = json.optJSONArray("decor");
            decor = new ArrayList<>();
            if (decorArray != null) {
                for (int i = 0; i < decorArray.length(); i++) decor.add(decorArray.getString(i));
            }
            JSONObject palette = json.getJSONObject("palette");
            primary = parseColor(palette.getString("primary"));
            secondary = parseColor(palette.getString("secondary"));
            accent = parseColor(palette.getString("accent"));
            neutral = parseColor(palette.getString("neutral"));
        }

        public boolean hasDecor(String name) {
            return decor.contains(name);
        }

        private static int parseColor(String hex) throws JSONException {
            try {
                return 0xFF000000 | Integer.parseInt(hex.substring(1), 16);
            } catch (RuntimeException error) {
                throw new JSONException("Invalid palette colour: " + hex);
            }
        }
    }

    private RenderManifest(JSONObject json) throws JSONException {
        version = json.getInt("version");
        if (version != SUPPORTED_VERSION) {
            throw new JSONException(
                "This app build renders manifest v" + SUPPORTED_VERSION + ", not v" + version);
        }
        attemptId = json.getString("attemptId");
        jobId = json.getString("jobId");
        requestId = json.getString("requestId");
        step = json.getString("step");
        stage = json.getString("stage");
        ratio = json.getString("ratio");
        width = json.getInt("width");
        height = json.getInt("height");
        fps = json.getInt("fps");
        sceneTransitionSeconds = json.optDouble("sceneTransitionSeconds", 0.2d);
        if (width <= 0 || height <= 0 || fps <= 0) throw new JSONException("Invalid canvas");

        sources = new ArrayList<>();
        JSONArray sourceArray = json.getJSONArray("sources");
        for (int i = 0; i < sourceArray.length(); i++) {
            sources.add(new Source(sourceArray.getJSONObject(i)));
        }

        scenes = new ArrayList<>();
        JSONArray sceneArray = json.getJSONArray("scenes");
        for (int i = 0; i < sceneArray.length(); i++) {
            scenes.add(new Scene(sceneArray.getJSONObject(i)));
        }

        inputVideoUrl = json.isNull("masterUrl") ? null : json.getString("masterUrl");
        voiceUrl = json.isNull("voiceUrl") ? null : json.getString("voiceUrl");
        musicUrl = json.isNull("musicUrl") ? null : json.getString("musicUrl");
        audio = new AudioSpec(json.getJSONObject("audio"));

        captions = new ArrayList<>();
        JSONArray captionArray = json.getJSONArray("captions");
        for (int i = 0; i < captionArray.length(); i++) {
            captions.add(new Caption(captionArray.getJSONObject(i)));
        }

        captionLanguages = new ArrayList<>();
        JSONArray languageArray = json.getJSONArray("captionLanguages");
        for (int i = 0; i < languageArray.length(); i++) {
            captionLanguages.add(languageArray.getString(i));
        }

        template = new Template(json.getJSONObject("template"));

        JSONObject output = json.getJSONObject("output");
        coverRequired = output.optBoolean("coverRequired", false);
        coverAtSeconds = output.optDouble("coverAtSeconds", 1d);
        maxOutputBytes = output.getLong("maxBytes");
        buildFromSources = json.optBoolean("buildFromSources", false);

        if (isMontage() && scenes.isEmpty()) {
            throw new JSONException("A montage manifest has no scenes");
        }
        if (!isMontage() && inputVideoUrl == null && scenes.isEmpty()) {
            throw new JSONException("Nothing to render from");
        }
        if (buildFromSources && (scenes.isEmpty() || sources.isEmpty())) {
            throw new JSONException("A render from sources needs its scenes and sources");
        }
        if (buildFromSources && !isMontage() && voiceUrl == null) {
            throw new JSONException("A render from sources needs the approved voice");
        }
    }

    public static RenderManifest parse(String json) throws JSONException {
        return new RenderManifest(new JSONObject(json));
    }

    public boolean isMontage() {
        return "montage".equals(stage);
    }

    public boolean isMaster() {
        return "master".equals(stage);
    }

    public boolean isFinal() {
        return "final".equals(stage);
    }

    /** Look a shot's source up; never null for a manifest that validated. */
    public Source sourceFor(Shot shot) throws JSONException {
        for (Source source : sources) {
            if (source.assetId.equals(shot.sourceAssetId)) return source;
        }
        throw new JSONException("A shot refers to a source the manifest does not carry");
    }

    /** Every shot in render order, flattened across scenes. */
    public List<Shot> flattenShots() {
        List<Shot> all = new ArrayList<>();
        for (Scene scene : scenes) all.addAll(scene.shots);
        return all;
    }

    /**
     * The dissolve length going INTO the shot at flat index `i`.
     *
     * Within a scene this is the scene's own `shotTransitionSeconds` (0 for a
     * "cut"); across a scene boundary it is the manifest's
     * `sceneTransitionSeconds`, which mirrors the server's `xfade` — unless the
     * scene's chosen transition INTO it is a cut, which the studio offers and
     * which is honoured here rather than dissolved over. The first shot of the
     * whole timeline has no dissolve.
     */
    public double dissolveBeforeFlatIndex(int flatIndex) {
        if (flatIndex <= 0) return 0d;
        int cursor = 0;
        for (Scene scene : scenes) {
            for (int i = 0; i < scene.shots.size(); i++) {
                if (cursor == flatIndex) {
                    if (i == 0) {
                        return "cut".equals(scene.transitionIn) ? 0d : sceneTransitionSeconds;
                    }
                    return "cut".equals(scene.transitionIn) ? 0d : scene.shotTransitionSeconds;
                }
                cursor++;
            }
        }
        return 0d;
    }

    /** Total picture length: every slot, end to end (dissolves borrow, never add). */
    public double pictureSeconds() {
        double total = 0d;
        for (Scene scene : scenes) {
            for (Shot shot : scene.shots) total += shot.durationSeconds;
        }
        return total;
    }

    /** The scene a flat shot index belongs to, for its transition flourish. */
    public Scene sceneForFlatIndex(int flatIndex) {
        int cursor = 0;
        for (Scene scene : scenes) {
            if (flatIndex < cursor + scene.shots.size()) return scene;
            cursor += scene.shots.size();
        }
        return scenes.isEmpty() ? null : scenes.get(scenes.size() - 1);
    }
}

File: ios/App/App/Render/RenderManifest.swift
import Foundation
import UIKit

/// The server's render manifest, decoded.
///
/// The Swift half of `src/lib/mobile/deviceRenderContract.ts`, and the exact
/// counterpart of Android's `RenderManifest.java`. Kept strict on purpose: a
/// field the manifest must carry is non-optional, so a manifest this build
/// half-understands fails to decode rather than producing a video that looks
/// nearly right.
struct RenderManifest: Decodable {

    static let supportedVersion = 4

    enum Stage: String, Decodable {
        /// Approved photos and clips become one silent intermediate.
        case montage
        /// That montage plus the mixed voice and ducked music.
        case master
        /// That master with the template and captions burned in, plus a cover.
        case final
    }

    enum Motion: String, Decodable {
        case kenBurnsIn = "ken_burns_in"
        case kenBurnsOut = "ken_burns_out"
        case panLeft = "pan_left"
        case panRight = "pan_right"
        case staticFrame = "static"
    }

    enum Transition: String, Decodable {
        case cut, fade, slide, zoom
    }

    struct Source: Decodable {
        let assetId: String
        /// "image" or "clip".
        let kind: String
        /// Exactly one of url / localId is present.
        let url: String?
        let localId: String?
        let mimeType: String
        let durationSeconds: Double?

        var isImage: Bool { kind == "image" }
    }

    struct Shot: Decodable {
        let sourceAssetId: String
        /// The slot this shot occupies before any dissolve overlap.
        let durationSeconds: Double
        let motion: Motion
        let trimStartSeconds: Double?
        let trimEndSeconds: Double?
        let focusX: Double
        let focusY: Double
        /// 0 = whole picture, 1 = fill the frame. Absent from an older plan,
        /// which means fill — what every render did before it existed.
        let frameZoom: Double?
        let playbackRate: Double

        var zoom: Double { min(1, max(0, frameZoom ?? 1)) }
    }

    struct Scene: Decodable {
        let sceneNumber: Int
        let transitionIn: Transition
        let shotTransitionSeconds: Double
        let assets: [Shot]
    }

    /// Every number the mixer needs; see `deviceRenderAudio.ts` for the why.
    struct AudioSpec: Decodable {
        let sourceClipAudio: Bool
        let musicSelected: Bool
        let sampleRate: Int
        let voiceLeadInSeconds: Double
        let musicBedVolume: Double
        let musicDuckRatio: Double
        let musicDuckThreshold: Double
        let musicDuckAttackMs: Double
        let musicDuckReleaseMs: Double
        let voiceTargetLufs: Double
        let voiceTruePeakDb: Double
        let voiceMaxGainDb: Double
        let mixLimit: Double
        let voiceDurationSeconds: Double?
    }

    struct Caption: Decodable {
        let startSeconds: Double
        let endSeconds: Double
        let textThai: String
        let textEnglish: String
        let textChinese: String

        func text(for language: String) -> String {
            switch language {
            case "th": return textThai
            case "en": return textEnglish
            case "zh": return textChinese
            default: return ""
            }
        }
    }

    struct Palette: Decodable {
        let primary: String
        let secondary: String
        let accent: String
        let neutral: String
    }

    struct Template: Decodable {
        let id: String
        let frame: String
        let canvas: String
        let decor: [String]
        let palette: Palette
    }

    struct Output: Decodable {
        let mimeType: String
        let videoCodec: String
        let audioCodec: String?
        let maxBytes: Int
        let coverRequired: Bool
        let coverAtSeconds: Double
    }

    let version: Int
    let attemptId: String
    let taskId: String
    let jobId: String
    let requestId: String
    let step: String
    let stage: Stage
    let travy: Bool
    let ratio: String
    let width: Int
    let height: Int
    let fps: Int
    let sceneTransitionSeconds: Double
    let sources: [Source]
    let scenes: [Scene]
    /// The montage for a master render, the master for a final render.
    let masterUrl: String?
    let voiceUrl: String?
    let musicUrl: String?
    let audio: AudioSpec
    let captions: [Caption]
    let captionLanguages: [String]
    let template: Template
    let output: Output
    /// Render this master or final straight from the originals in ONE export
    /// (plugin version 6+). Absent on older manifests, hence optional.
    let buildFromSources: Bool?

    var composesFromSources: Bool { buildFromSources ?? false }

    static func parse(_ json: String) throws -> RenderManifest {
        guard let data = json.data(using: .utf8) else {
            throw RenderError.manifest("The render manifest is not valid UTF-8")
        }
        let manifest = try JSONDecoder().decode(RenderManifest.self, from: data)
        guard manifest.version == supportedVersion else {
            throw RenderError.manifest(
                "This app build renders manifest v\(supportedVersion), not v\(manifest.version)")
        }
        guard manifest.width > 0, manifest.height > 0, manifest.fps > 0 else {
            throw RenderError.manifest("The manifest has an invalid canvas")
        }
        if manifest.stage == .montage && manifest.scenes.isEmpty {
            throw RenderError.manifest("A montage manifest has no scenes")
        }
        if manifest.stage != .montage && manifest.masterUrl == nil && manifest.scenes.isEmpty {
            throw RenderError.manifest("Nothing to render from")
        }
        if manifest.composesFromSources && (manifest.scenes.isEmpty || manifest.sources.isEmpty) {
            throw RenderError.manifest("A render from sources needs its scenes and sources")
        }
        if manifest.composesFromSources && manifest.stage != .montage && manifest.voiceUrl == nil {
            throw RenderError.manifest("A render from sources needs the approved voice")
        }
        return manifest
    }

    var canvasSize: CGSize { CGSize(width: width, height: height) }

    /// Total picture length: every slot end to end (dissolves borrow, never add).
    var pictureSeconds: Double {
        scenes.reduce(0) { total, scene in total + scene.assets.reduce(0) { $0 + $1.durationSeconds } }
    }

    func source(for shot: Shot) throws -> Source {
        guard let source = sources.first(where: { $0.assetId == shot.sourceAssetId }) else {
            throw RenderError.manifest("A shot refers to a source the manifest does not carry")
        }
        return source
    }

    /// Every shot in render order, flattened across scenes.
    var flattenedShots: [Shot] { scenes.flatMap { $0.assets } }

    /// The scene a flat shot index belongs to, for its transition flourish.
    func scene(forFlatIndex index: Int) -> Scene? {
        var cursor = 0
        for scene in scenes {
            if index < cursor + scene.assets.count { return scene }
            cursor += scene.assets.count
        }
        return scenes.last
    }

    /// The dissolve going INTO the shot at flat index `index`.
    ///
    /// Within a scene this is the scene's own `shotTransitionSeconds` (0 for a
    /// cut); across a scene boundary it is `sceneTransitionSeconds`, mirroring
    /// the server's `xfade` — unless the scene's chosen transition into it is a
    /// cut, which the studio offers and which is honoured rather than dissolved
    /// over. The first shot of the timeline has no dissolve.
    func dissolve(beforeFlatIndex index: Int) -> Double {
        guard index > 0 else { return 0 }
        var cursor = 0
        for scene in scenes {
            for i in 0..<scene.assets.count {
                if cursor == index {
                    if i == 0 { return scene.transitionIn == .cut ? 0 : sceneTransitionSeconds }
                    return scene.transitionIn == .cut ? 0 : scene.shotTransitionSeconds
                }
                cursor += 1
            }
        }
        return 0
    }
}

/// One error type for the whole render path, so a failure that reaches the web
/// layer always carries a sentence a tester can act on rather than an NSError
/// code they have to look up.
enum RenderError: LocalizedError {
    case manifest(String)
    case missingInput(String)
    case export(String)
    case audio(String)
    case cancelled

    var errorDescription: String? {
        switch self {
        case .manifest(let message): return message
        case .missingInput(let message): return "This render needs \(message), which is not available"
        case .export(let message): return message
        case .audio(let message): return message
        case .cancelled: return "Render cancelled"
        }
    }
}
